staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go

```diff
@@ -178,19 +174,13 @@ func (c *threadSafeMap) Index(indexName string, obj interface{}) ([]interface{},
 
 // ByIndex returns a list of the items whose indexed values in the given index include the given indexed value
 func (c *threadSafeMap) ByIndex(indexName, indexedValue string) ([]interface{}, error) {
-	c.lock.RLock()
-	defer c.lock.RUnlock()
-
-	indexFunc := c.indexers[indexName]
-	if indexFunc == nil {
-		return nil, fmt.Errorf("Index with name %s does not exist", indexName)
+	keys, err := c.IndexKeys(indexName, indexedValue)
+	if err != nil {
+		return nil, err
 	}
 
-	index := c.indices[indexName]
-
-	set := index[indexedValue]
-	list := make([]interface{}, 0, set.Len())
-	for key := range set {
+	list := make([]interface{}, 0, len(keys))
+	for _, key := range keys {
 		list = append(list, c.items[key])
 	}
```

Comment:
I'm not sure this is correct:
- In the original implementation a read lock is held while reading values from the index, hence making it thread safe.
- In the change done, the read lock is going to be held *only* while `IndexKeys` is executed, post which operations are not under a lock which most likely will lead to issues.

Reply:
you are right, i will revert the change